src/v/cloud_storage/remote.cc

```diff
@@ -0,0 +1,494 @@
+            // exist)
+            blog(
+              "Accessing {}, unexpected REST API error \"{}\" detected, code: "
+              "{}, request_id: {}, resource: {}",
+              bucket,
+              err.message(),
+              err.code_string(),
+              err.request_id(),
+              err.resource());
+            result = error_outcome::fail;
+        }
+    } catch (const std::system_error& cerr) {
+        // The system_error is type erased and not convenient for selective
+        // handling. The following errors should be retried:
+        // - connection refused, timed out or reset by peer
+        // - network temporary unavailable
+        // Shouldn't be retried
+        // - any filesystem error
+        // - broken-pipe
+        // - any other network error (no memory, bad socket, etc)
+        if (auto code = cerr.code(); code.value() != ECONNREFUSED
+                                     && code.value() != ENETUNREACH
+                                     && code.value() != ECONNRESET) {
+            blog("System error {}", cerr);
+            result = error_outcome::fail;
+        }
+        blog("System error susceptible for retry {}", cerr.what());
+    } catch (ss::timed_out_error& terr) {
+        // This should happen when the connection pool was disconnected
+        // from the S3 endpoint and subsequent connection attmpts failed.
+        blog("Connection timeout {}", terr.what());
+    } catch (const boost::system::system_error& err) {
+        if (auto code = err.code();
+            string_switch<bool>(code.category().name())
+              .match("beast.http", false)
+              .default_match(true)
+            || code.value()
+                 != static_cast<int>(boost::beast::http::error::short_read)) {
+            blog("Connection failed {}", err.what());
+            result = error_outcome::fail;
+        } else {
+            // This is a short read error that can be caused by the abrupt TLS
+            // shutdown. The content of the received buffer is discarded in this
+            // case and http client receives an empty buffer.
+            blog("Slow Read detected, retrying {}", err.what());
+        }
+    }
+    return result;
+}
+
+remote::remote(
+  s3_connection_limit limit,
+  const s3::configuration& conf,
+  service_probe& probe)
+  : _pool(limit(), conf)
+  , _probe(probe) {}
+
+ss::future<> remote::start() { return ss::now(); }
+
+ss::future<> remote::stop() {
+    _as.request_abort();
+    co_await _pool.stop();
+    co_return co_await _gate.close();
+}
+
+ss::future<download_result> remote::download_manifest(
+  const s3::bucket_name& bucket,
+  base_manifest& manifest,
+  ss::lowres_clock::duration timeout) {
+    gate_guard guard{_gate};
+    burst_logger blog;
+    // NOTE: all log messages repeatedly generated by the retries
+    // should go to he burst_log, everything else is logged as ususal
+    backoff_helper backoff{timeout};
```

Comment:
Maybe this is useful: https://github.com/vectorizedio/redpanda/blob/082a93c03de23a52326e28b7f51ac413b854cb75/src/v/utils/retry.h